lib/core/src/display/shape/text/cursor.rs

```diff
@@ -63,15 +72,165 @@ impl Cursor {
+            line   : line_index,
+        }
+    }
+
+    /// Get cursor position at end of given line.
+    pub fn line_end(&self, line_index:usize) -> CharPosition {
+        CharPosition{
+            line   : line_index,
+            column : self.content.lines[line_index].len(),
+        }
+    }
+
+    /// Get cursor position at begin of whole content
+    pub fn content_begin(&self) -> CharPosition {
+        CharPosition {
+            column : 0,
+            line   : 0,
+        }
+    }
+
+    /// Get cursor position at end of whole content
+    pub fn content_end(&self) -> CharPosition {
+        CharPosition {
+            column : self.content.lines.last().unwrap().len(),
+            line   : self.content.lines.len() - 1,
+        }
+    }
+
+    /// Get cursor position for the next char from given position. Returns none if at end of
+    /// whole document.
+    pub fn next_char(&self, position:&CharPosition) -> Option<CharPosition> {
+        let current_line = &self.content.lines[position.line];
+        let next_column  = Some(position.column + 1).filter(|c| *c <= current_line.len());
+        let next_line    = Some(position.line + 1)  .filter(|l| *l < self.content.lines.len());
+        match (next_column,next_line) {
+            (None         , None      ) => None,
+            (None         , Some(line)) => Some(self.line_begin(line)),
+            (Some(column) , _         ) => Some(CharPosition{column, ..*position})
+        }
+    }
+
+    /// Get cursor position for the previous char from given position. Returns none if at begin of
+    /// whole document.
+    pub fn prev_char(&self, position:&CharPosition) -> Option<CharPosition> {
+        let prev_column = position.column.checked_sub(1);
+        let prev_line   = position.line.checked_sub(1);
+        match (prev_column,prev_line) {
+            (None         , None      ) => None,
+            (None         , Some(line)) => Some(self.line_end(line)),
+            (Some(column) , _         ) => Some(CharPosition{column, ..*position})
+        }
+    }
+
+    /// Get cursor position one line above the given position, such the new x coordinate of
+    /// displayed cursor on the screen will be nearest the current value.
+    pub fn line_up(&mut self, position:&CharPosition) -> Option<CharPosition> {
+        let prev_line = position.line.checked_sub(1);
+        prev_line.map(|line| self.near_same_x_in_another_line(position,line))
+    }
+
+    /// Get cursor position one line behind the given position, such the new x coordinate of
+    /// displayed cursor on the screen will be nearest the current value.
+    pub fn line_down(&mut self, position:&CharPosition) -> Option<CharPosition> {
+        let next_line = Some(position.line + 1).filter(|l| *l < self.content.lines.len());
+        next_line.map(|line| self.near_same_x_in_another_line(position,line))
+    }
+
+    fn new_position(&mut self, position:CharPosition, step:&Step) -> CharPosition {
```

Comment:
I know this is private function, but a comment of what it does and what it is used for would be really helpful here.